src/librustc_typeck/collect.rs

```diff
@@ -1410,118 +1377,252 @@ fn convert_trait_predicates<'a, 'tcx>(ccx: &CrateCtxt<'a, 'tcx>, it: &hir::Item)
     }
 }
 
-fn type_scheme_of_def_id<'a,'tcx>(ccx: &CrateCtxt<'a,'tcx>,
-                                  def_id: DefId)
-                                  -> ty::TypeScheme<'tcx>
-{
-    if let Some(node_id) = ccx.tcx.map.as_local_node_id(def_id) {
-        match ccx.tcx.map.find(node_id) {
-            Some(hir_map::NodeItem(item)) => {
-                type_scheme_of_item(ccx, &item)
+fn generics_of_def_id<'a, 'tcx>(ccx: &CrateCtxt<'a, 'tcx>,
+                                def_id: DefId)
+                                -> &'tcx ty::Generics<'tcx> {
+    let tcx = ccx.tcx;
+    let node_id = if let Some(id) = tcx.map.as_local_node_id(def_id) {
+        id
+    } else {
+        return tcx.lookup_generics(def_id);
+    };
+    tcx.generics.memoize(def_id, || {
+        use rustc::hir::map::*;
+        use rustc::hir::*;
+
+        let mut opt_self = None;
+        let mut base_def_id = None;
+        let mut allow_defaults = false;
+
+        let no_generics = hir::Generics::empty();
+        let (space, ast_generics) = match tcx.map.get(node_id) {
+            NodeImplItem(&ImplItem { node: ImplItemKind::Method(ref sig, _), .. }) |
+            NodeTraitItem(&TraitItem { node: MethodTraitItem(ref sig, _), .. }) => {
+                let parent_id = tcx.map.get_parent(node_id);
+                base_def_id = Some(tcx.map.local_def_id(parent_id));
+                (FnSpace, &sig.generics)
             }
-            Some(hir_map::NodeForeignItem(foreign_item)) => {
-                let abi = ccx.tcx.map.get_foreign_abi(node_id);
-                type_scheme_of_foreign_item(ccx, &foreign_item, abi)
+
+            NodeImplItem(_) |
+            NodeTraitItem(_) => {
+                let parent_id = tcx.map.get_parent(node_id);
+                base_def_id = Some(tcx.map.local_def_id(parent_id));
+                (FnSpace, &no_generics)
             }
-            x => {
-                bug!("unexpected sort of node in get_item_type_scheme(): {:?}",
-                     x);
+
+            NodeItem(item) => {
+                match item.node {
+                    ItemFn(_, _, _, _, ref generics, _) => (FnSpace, generics),
```

Comment:
Nit: I'd prefer a newline before the `=>` here and in the next arm, so that these cases are clearly distinct from the three that are combined with `|` below.